Document approach to using private utils

### What is the problem you're trying to solve?

I'm writting a custom stylelint rule in TS,  as mentioned in the document:
[doc link](https://stylelint.io/developer-guide/rules#utility-functions)
> Stylelint has [utility functions](https://github.com/stylelint/stylelint/tree/main/lib/utils) that are used in existing rules and might prove useful to you, as well. Please look through those so that you know what's available. (And if you have a new function that you think might prove generally helpful, let's add it to the list!).

I want to use some utils functions to help me implement the rule, like:
`example-rule.ts`
```typescript
import postcss from 'postcss';
import stylelint, { utils } from 'stylelint';

import eachDeclarationBlock from 'stylelint/lib/utils/eachDeclarationBlock';
import isKeyframeSelector from 'stylelint/lib/utils/isKeyframeSelector';
import isStandardSyntaxRule from 'stylelint/lib/utils/isStandardSyntaxRule';
import isStandardSyntaxSelector from 'stylelint/lib/utils/isStandardSyntaxSelector';
import { isDeclaration } from 'stylelint/lib/utils/typeGuards';
import { isBoolean, isRegExp, isString } from 'stylelint/lib/utils/validateTypes';

const rule: stylelint.Rule = (primary, secondaryOptions) => {
  return (root, result) => {
    // use utils functions bellow
    eachDeclarationBlock(root, eachDecl => {
      eachDecl(decl => {
        isDeclaration(decl);
      });
    });

    isDeclaration(root);
    isKeyframeSelector('');
    isStandardSyntaxSelector('');
    isStandardSyntaxRule(null! as postcss.Rule);
    isBoolean(null! as unknown);
    isRegExp(null! as unknown);
    isString(null! as unknown);

    utils.report(null! as stylelint.Problem);
  }
};

rule.ruleName = '';
rule.messages = null!;

```

But TS complained that  it could not find the declaration file:
```text
example-rule.ts:4:34 - error TS7016: Could not find a declaration file for module 'stylelint/lib/utils/eachDeclarationBlock'. 'stylelint-plugin-example/node_modules/stylelint/lib/utils/eachDeclarationBlock.js' implicitly has an 'any' type.
  If the 'stylelint' package actually exposes this module, try adding a new declaration (.d.ts) file containing `declare module 'stylelint/lib/utils/eachDeclarationBlock';`

4 import eachDeclarationBlock from 'stylelint/lib/utils/eachDeclarationBlock';

... and more error like above
```

### What solution would you like to see?

I can write my rule like the code above without any type errors and get type hints.

for now, I do this by modifying the local `stylelint` repository and linking to my `custom-rule` repository via "npm link", and it works well. 
I create a new tsconfig file in my local `stylelint` project to emit declaration files:
`tsconfig.emit.json`
```json
{
	"extends": "./tsconfig",
	"include": ["lib/"],
	"compilerOptions": {
		"declaration": true,
		"noEmit": false,
		"emitDeclarationOnly": true,
		"noEmitOnError": true
	}
}
```
then, i run `tsc -p tsconfig.emit.json` to emit declaration files, and link `stylelint` to global by `npm link`.
After that, I switch to my `custom-rule` project and run `npm link stylelint`,  everything is fine.

--- a/docs/developer-guide/plugins.md
+++ b/docs/developer-guide/plugins.md
@@ -10,6 +10,7 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
 - tests
 - docs
 - metadata
+- construct-specific parsers
 
 ## The anatomy of a plugin
 
@@ -216,6 +217,8 @@ Alternatively, if you don't want to use Jest you'll find more testing tool in [A
 
 Stylelint exposes some useful utilities.
 
+You're also welcome to copy any of the [internal utils](https://github.com/stylelint/stylelint/tree/main/lib/utils) into your plugin. You should not `require` or `import` them directly, as they are not part of the public API and may change or be removed without warning.
+
 ### `stylelint.utils.report`
 
 Adds problems from your plugin to the list of problems that Stylelint will report to the user.
@@ -310,14 +313,6 @@ module.exports = stylelint.createPlugin(ruleName, (expectation) => {
 ## Allow primary option arrays
 
 If your plugin can accept an array as its primary option, you must designate this by setting the property `primaryOptionArray = true` on your rule function. For more information, check out the ["Working on rules"](rules.md) doc.
-
-## External helper modules
-
-In addition to the standard parsers mentioned in the ["Working on rules"](rules.md) doc, there are other external modules used within Stylelint that we recommend using. These include:
-
-- [postcss-resolve-nested-selector](https://github.com/davidtheclark/postcss-resolve-nested-selector): given a (nested) selector in a PostCSS AST, return an array of resolved selectors.
-
-Have a look through [Stylelint's internal utils](https://github.com/stylelint/stylelint/tree/main/lib/utils) and if you come across one that you need in your plugin, then please consider helping us extract it into an external module.
 
 ## Peer dependencies
 

--- a/docs/developer-guide/rules.md
+++ b/docs/developer-guide/rules.md
@@ -97,6 +97,7 @@ Use `node.raws` instead of `node.raw()` when accessing raw strings from the [Pos
 
 Depending on the rule, we also recommend using:
 
+- [media-query-list-parser](https://www.npmjs.com/package/@csstools/media-query-list-parser)
 - [postcss-value-parser](https://github.com/TrySound/postcss-value-parser)
 - [postcss-selector-parser](https://github.com/postcss/postcss-selector-parser)
 

